src/Raven.Server/Documents/TransactionOperationsMerger.cs

```diff
@@ -110,6 +162,178 @@ private static void ThrowTxMergerWasDisposed()
             throw new ObjectDisposedException("Transaction Merger");
         }
 
+
+        public abstract class RecordingState : IDisposable
+        {
+            public abstract void Record(DocumentsOperationContext context, MergedTransactionCommand cmd);
+
+            public abstract void Record(DocumentsOperationContext context, TxInstruction tx, bool doRecord = true);
+
+            public abstract void Prepare(ref RecordingState state);
+
+            private static BlittableJsonReaderObject SerializeRecordingCommandDetails(
+                JsonOperationContext context,
+                RecordingDetails commandDetails)
+            {
+                using (var writer = new BlittableJsonWriter(context))
+                {
+                    var jsonSerializer = GetJsonSerializer();
+
+                    jsonSerializer.Serialize(writer, commandDetails);
+                    writer.FinalizeDocument();
+
+                    return writer.CreateReader();
+                }
+            }
+
+            protected class EnabledRecordingState : RecordingState
+            {
+                private readonly TransactionOperationsMerger _txOpMerger;
+                private Stream _recordingStream;
+                private bool _isDisposed = false;
+
+                public EnabledRecordingState(TransactionOperationsMerger txOpMerger)
+                {
+                    _txOpMerger = txOpMerger;
+                }
+
+                public void Init(string filePath)
+                {
+                    var recordingFileStream = new FileStream(filePath, FileMode.Create);
```

Comment:
If we provide a bad file path (`f:\foo`, no permissions, file already exists, etc). What will happen?

Reply:
I choose to move the ownership of the record stream to the TxMerger and create the file while the start record request so the error will be arrived from there.
As I implemented it now the record will overwrite the existing file.
I can think about three ways to handle it:
1. To overwrite
2. To return error
3. To add an extension (maybe current time) that differentiate it from the existing